cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -149,9 +150,22 @@ class ParquetScanTaskIterator {
   static Status InferColumnProjection(const parquet::FileMetaData& metadata,
                                       const std::shared_ptr<ScanOptions>& options,
                                       std::vector<int>* out) {
-    // TODO(fsaintjacques): Compute intersection _and_ validity
-    *out = internal::Iota(metadata.num_columns());
+    if (options->projector == nullptr || options->schema == nullptr) {
+      *out = internal::Iota(metadata.num_columns());
+      return Status::OK();
+    }
 
+    // union fields in expression and fields in projector
+    auto projector_schema = options->projector->schema();
+    auto filter_schema =
+        SchemaFromColumnNames(options->schema, FieldsInExpression(options->filter));
+    ARROW_ASSIGN_OR_RAISE(auto merged, MergeSchemas({projector_schema, filter_schema}));
```

Comment:
This function tries to apply an optimization, if it can't it shouldn't error but simply fallback to the slow-path, e.g. `*out = internal::Iota(metadata.num_columns());`.